README.md
# WiFiCart
WiFi Cart for the Commodore 64 with MicroView


Firmware/changelog.txt
Version 0.13 - WiFiModem-Menu-013.hex / WiFiModem-Hayes-013.hex
---------------------------------------------------------------
Dec 7, 2017

New:
-Fix for connecting to Quantum Link.  See documentation for new
 instructions.

Hayes:
-New AT*Bn option to hard code baud rate to 1200, 2400 or 9600.


Version 0.12 - WiFiModem-Menu-012.hex / WiFiModem-Hayes-012.hex
---------------------------------------------------------------
Feb 24, 2016

Hayes:
-CommodoreServer 38,400 baud improvements.
-AT&S0/1/2 DSR fix.
-Fix for entering S2 register on command line that also includes * settings.


Version 0.11 - WiFiModem-Menu-011.hex / WiFiModem-Hayes-011.hex
---------------------------------------------------------------
Feb 12, 2016

New:
-Improved telnet detection.  Eliminates delay when connecting to the modem
 from remote.
-File transfers over telnet now works by detecting telnet binary mode.
 Previously only raw transfers were working reliably.

Menu:
-After accepting a call, modem waits for another call instead of going back
 to the menu.  Pressing a key will abort while waiting for a call.

Hayes:
-Fix for ATS2 command register for changing escape character.  The s register
 could be changed, but it had no effect.
-For ATS2 escape, a value of 128-255 now disables the escape code.
-PETSCII response fixes.
-OK/ERROR response fixes.
-Added AT&S0/1/2 to set DSR option.  
 &S0 = DSR always on
 &S1 = DSR will become active when carrier has been detected and
       inactive when carrier dropped.  Similar to DCD.
 &S2 = Disable DSR line.  Set this when using UP9600.
-Fix for SSID / passphrase not being saved after being set.
-Replaced 'CONNECTING...' message with a custom message that can be set using
 AT*M=.  The message will be displayed after a remote user connects (incoming 
 connections only).
-DCD now defaults to inverted.  In Novaterm, select 'Normal' for carrier.
-AT command line can now contain non-AT command text before AT.  For example:
 abcATE0 is now accepted.  Required for some BBSs.
-Fix for CommodoreServer at 38,400 baud not working if the modem was
 previously set to 1200 baud.


Version 0.10 - WiFiModem-Menu-010.hex / WiFiModem-Hayes-010.hex
---------------------------------------------------------------
Jan 17, 2016

New:

Hayes:
-Sending AT commands in uppercase PETSCII is now supported.  Response will 
 also be in PETSCII uppercase (0xc1-0xda) as seen on real modems.
-Removed PETSCII translation for text input and displayed text to save space.
 Use ASCII mode for setting the SSID / passphrase.
-Fix for modem not always going off-hook with ath1.
-ath0/1 now prints OK instead of NO CARRIER.
-auto-answer now prints RING before connecting.
-CommodoreServer.com 38,400 baud support.  See manual for details.
-ATS99 command register to disable AT ERROR responses.  Should only be used 
 when running a program that uses AT commands that are not supported by the
 modem and does not allow for the AT initialization string to be changed.
-Removed ATM0-3 support which did nothing but was previously added to prevent 
 an AT command error with BBSs that send it in the init string.  Use ATS99
 instead.
-Added AT*M= command to set a message that will be displayed after a remote 
 user connects (incoming connections).


Version 0.09 - WiFiModem-Menu-009.hex / WiFiModem-Hayes-009.hex
---------------------------------------------------------------
Jan 10, 2016

New:
-Fix for issue with C64 to modem flow control
-Improved telnet detection

Hayes:
-Can now escape to AT command mode after connecting to a host using +++.
-Guard time added to +++ (must wait 1 second before issuing +++)
-ATS2 command register for changing escape character.  Default is ATS2=43 (+).
-ATM0-3 support.  Does nothing but won't cause an AT command error for BBSs 
 that send it in init string.
-ATX0-4 support:
   X0 = 0-4
   X1 = 0-5, 10+
   X2 = 0-6, 10+
   X3 = 0-5, 7, 10+
   X4 = 0-7, 10+

   0 - OK
   1 - CONNECT
   2 - RING
   3 - NO CARRIER
   4 - ERROR
   5 - CONNECT 1200
   6 - NO DIALTONE	(DNS failure)
   7 - BUSY		(Connection failure)
   8 - NO ANSWER	(not implemented)
   10 - CONNECT 2400
   11 - CONNECT 4800
   .
   .
-1200 baud support for inbound connections.
-ATH1 to take modem off-hook (WiFly goes to sleep) to prevent incoming connections.
-Default phone book entries when clearing have been removed.
-AT&SSID= changed to AT*S=
-AT&PASS= changed to AT*P=
-AT&KEY= changed to AT*K=

Menu:
-Can now abort waiting for inbound connection by pressing any key.
-Option to clear phone book
-Improved character input routines in menus


Version 0.08 - WiFiModem-Menu-008.hex / WiFiModem-Hayes-008.hex
---------------------------------------------------------------
Dec 28, 2015

New:
-Q-Link now works at 1200 baud.  See the Commodore WiFi Modem Users Guide.pdf document for instructions.
-Dialing 555-1212 will connect to lyonlabs.org:5190
-Dialing 555-1213 will connect to q-link.org:5190
-No longer forces flow control at 1200 baud

Hayes:
-fixed bug with 1200 baud commands not working when flow control is off
-atz now loads the saved settings instead of factory defaults.  Use at&f for factory defaults.
-ati now shows both current and saved settings.


Version 0.07 - WiFiModem-Menu-007.hex / WiFiModem-Hayes-007.hex
---------------------------------------------------------------
Dec 18, 2015

New:
-Add support for DCD always on / follow remote carrier to menu and Hayes (at&c0/1)
-Microview wordwrap fixed
-Fix for answering calls at > 2400 baud
-Incoming connections now negotiate telnet parameters to prevent junk from being displayed during initial connection
-Incoming connections now support bi-directional flow control.

Menu:

-Phone book added.

MicroView/WiFiModem-Hayes/ModemBase.cpp
/***********************************************
HAYESDUINO PROJECT - COPYRIGHT 2013, PAYTON BYRD

This version modded by Leif Bloomquist for use on the Wi-Fi Modem!

Project homepage: http://hayesduino.codeplex.com
License: http://hayesduino.codeplex.com/license
***********************************************/

#include <WiFlyHQ.h>
#include "..\..\..\hayesduino\ModemBase.h"

int ModemBase::available(void) 
{ 
	if(_serial) return _serial->available(); 
	return 0; 
}

int ModemBase::peek(void) 
{
	if(_serial) return _serial->peek(); 
    return 0; 
}

int ModemBase::read(void) 
{
	if(_serial) return _serial->read(); 
    return 0; 
}

void ModemBase::flush(void) 
{
	if(_serial) _serial->flush(); 
}
size_t ModemBase::write(uint8_t var) 
{
	if(_serial) return _serial->write(var); 
	return 0; 
}

void ModemBase::resetCommandBuffer()
{
	memset(_commandBuffer, 0, strlen(_commandBuffer));
}

ModemBase::ModemBase()
{
	pinMode(RI , OUTPUT);
	pinMode(CTS, OUTPUT);	
	pinMode(DSR, OUTPUT);
	pinMode(DTR, OUTPUT);
	pinMode(DCD, INPUT);
	pinMode(RTS, INPUT);
	
	digitalWrite(RI, LOW);
	digitalWrite(CTS, HIGH);
	digitalWrite(DSR, HIGH);
	digitalWrite(DTR, toggleCarrier(false));
}

void ModemBase::begin(Stream *serial, WiFly *wifly, void(*onDialoutHandler)(char*), void(*onDisconnectHandler)())
{
	_serial = serial;
	_wifly = wifly;
	onDialout = onDialoutHandler;
	onDisconnect = onDisconnectHandler;
	
	_escapeCount = 0;

	_isCommandMode = true;
	_isConnected = false;
	_isRinging = false;

	loadDefaults();

	digitalWrite(DTR, toggleCarrier(false));    // These look weird - compare with above and RS232 spec
	digitalWrite(RTS, HIGH);
	digitalWrite(RI, LOW);

	resetCommandBuffer();
}

void ModemBase::loadDefaults(void)
{
	_echoOn = true;
	_verboseResponses = true;
	_quietMode = false;
	_S0_autoAnswer = false;
}

uint32_t ModemBase::getBaudRate(void)
{
	return _baudRate;
}

void ModemBase::setDefaultBaud(uint32_t baudRate)
{
	_baudRate = baudRate;
}

void ModemBase::setDcdInverted(char isInverted)
{
	_isDcdInverted = isInverted;
}

bool ModemBase::getDcdInverted(void)
{
	return _isDcdInverted == 1;
}

bool ModemBase::getIsConnected(void)
{
	return _isConnected;
}

bool ModemBase::getIsRinging(void)
{
	return _isRinging;
}


bool ModemBase::getIsCommandMode(void)
{
	return _isCommandMode;
}

int ModemBase::toggleCarrier(boolean isHigh)
{

	int result = 0; //_isDcdInverted ? (isHigh ? LOW : HIGH) : (isHigh ? HIGH : LOW);
	switch(_isDcdInverted)
	{
	case 0: result = (int)(!isHigh); break;
	case 1: result = (int)(isHigh); break;
	case 2: result = LOW;
	}
	//digitalWrite(STATUS_LED, result);
	return result;
}

void ModemBase::disconnect()
{
	_isCommandMode = true;
	_isConnected = false;
	_isRinging = false;

	// TODO - According to http://totse2.net/totse/en/technology/telecommunications/trm.html
	//		  The BBS should detect <CR><LF>NO CARRIER<CR><LF> as a dropped carrier sequences.
	//		  DMBBS does not honor this and so I haven't sucessfully tested it, thus it's commented out.

	//delay(100);
	//print('\r'); 
	//print((char)_S4_lfCharacter);
	//print(F("NO CARRIER"));
	//print('\r'); 
	//print((char)_S4_lfCharacter);

	digitalWrite(RTS, LOW);
	digitalWrite(DTR, toggleCarrier(false));
	onDisconnect();
}

void ModemBase::processCommandBuffer()
{
	for (int i=0; i < strlen(_commandBuffer); ++i)
	{
		_commandBuffer[i] = toupper(_commandBuffer[i]);
	}

	if (strcmp(_commandBuffer, ("AT/")) == 0)
	{
		strcpy(_commandBuffer, _lastCommandBuffer);
	}
	
	if (strcmp(_commandBuffer, ("ATZ")) == 0)
	{
		loadDefaults();
		printOK();
	}
	else if (strcmp(_commandBuffer, ("ATI")) == 0)
	{
		ShowInfo();
		printOK();
	}
	else if (strcmp(_commandBuffer, ("AT&F")) == 0)
	{
		if(strcmp(_lastCommandBuffer, ("AT&F")) == 0)
		{
			loadDefaults();
			printOK();
		}
		else
		{
			_serial->println(F("send command again to verify."));
		}
	}
	else if(strcmp(_commandBuffer, ("ATA")) == 0)
	{
		answer();
	}
	else if (strcmp(_commandBuffer, ("ATD")) == 0 || strcmp(_commandBuffer, ("ATO")) == 0)
	{
		if (_isConnected)
		{
			_isCommandMode = false;
		}
		else
		{
			printError();
		}
	}
	else if(
		strncmp(_commandBuffer, ("ATDT "), 5) == 0 ||
		strncmp(_commandBuffer, ("ATDP "), 5) == 0 ||
		strncmp(_commandBuffer, ("ATD "), 4) == 0
		)
	{
		onDialout(strstr(_commandBuffer, " ") + 1);	
		resetCommandBuffer();  // This avoids port# string fragments on subsequent calls
	}
	else if (strncmp(_commandBuffer, ("ATDT"), 4) == 0)
	{
		onDialout(strstr(_commandBuffer, "ATDT") + 4);
		resetCommandBuffer();  // This avoids port# string fragments on subsequent calls
	}
	else if ((strcmp(_commandBuffer, ("ATH0")) == 0 || strcmp(_commandBuffer, ("ATH")) == 0))
	{
		disconnect();
	}
	else if(strncmp(_commandBuffer, ("AT"), 2) == 0)
	{
		if(strstr(_commandBuffer, ("E0")) != NULL)
		{
			_echoOn = false;
		}

		if(strstr(_commandBuffer, ("E1")) != NULL)
		{
			_echoOn = true;
		}

		if(strstr(_commandBuffer, ("Q0")) != NULL)
		{
			_verboseResponses = false;
			_quietMode = false;
		}

		if(strstr(_commandBuffer, ("Q1")) != NULL)
		{
			_quietMode = true;
		}

		if(strstr(_commandBuffer, ("V0")) != NULL)
		{
			_verboseResponses = false;
		}

		if(strstr(_commandBuffer, ("V1")) != NULL)
		{
			_verboseResponses = true;
		}

		if(strstr(_commandBuffer, ("X0")) != NULL)
		{
			// TODO
		}
		if(strstr(_commandBuffer, ("X1")) != NULL)
		{
			// TODO
		}

		char *currentS;
		char temp[100];

		int offset = 0;
		if((currentS = strstr(_commandBuffer, ("S0="))) != NULL)
		{
			offset = 3;
			while(currentS[offset] != '\0' && isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 3, offset - 3);
			_S0_autoAnswer = atoi(temp);
		}
/*
		if((currentS = strstr(_commandBuffer, ("S1="))) != NULL)
		{
			offset =3;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 3, offset - 3);
			_S1_ringCounter = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S1=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S2="))) != NULL)
		{
			offset =3;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 3, offset - 3);
			_S2_escapeCharacter = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S2=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S3="))) != NULL)
		{
			offset =3;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 3, offset - 3);
			_S3_crCharacter = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S3=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S4="))) != NULL)
		{
			offset =3;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 3, offset - 3);
			_S4_lfCharacter = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S4=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S5="))) != NULL)
		{
			offset =3;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 3, offset - 3);
			_S5_bsCharacter = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S5=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S6="))) != NULL)
		{
			offset =3;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 3, offset - 3);
			_S6_waitBlindDial = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S6=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S7="))) != NULL)
		{
			offset =3;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 3, offset - 3);
			_S7_waitForCarrier = atoi(temp);
#if DEBUG == 1
			lggr.print(("S7=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S8="))) != NULL)
		{
			offset =3;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 3, offset - 3);
			_S8_pauseForComma = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S8=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S9="))) != NULL)
		{
			offset =3;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 3, offset - 3);
			_S9_cdResponseTime = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S9=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S10="))) != NULL)
		{
			offset =4;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 4, offset - 4);
			_S10_delayHangup = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S10=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S11="))) != NULL)
		{
			offset =4;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 4, offset - 4);
			_S11_dtmf = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S11=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S12="))) != NULL)
		{
			offset =4;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 4, offset - 4);
			_S12_escGuardTime = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S12=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S18="))) != NULL)
		{
			offset =4;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 4, offset - 4);
			_S18_testTimer = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S18=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S25="))) != NULL)
		{
			offset =4;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 4, offset - 4);
			_S25_delayDTR = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S25=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S26="))) != NULL)
		{
			offset =4;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 4, offset - 4);
			_S26_delayRTS2CTS = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S26=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S30="))) != NULL)
		{
			offset =4;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 4, offset - 4);
			_S30_inactivityTimer = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S30=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S37="))) != NULL)
		{
			offset =4;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 4, offset - 4);
			_S37_lineSpeed = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S37=")); lggr.println(temp);
#endif

			EEPROM.write(S37_ADDRESS, _S37_lineSpeed);

			setLineSpeed();

#if DEBUG == 1
			lggr.print(F("Set Baud Rate: ")); lggr.println(_baudRate);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S38="))) != NULL)
		{
			offset =4;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 4, offset - 4);
			_S38_delayForced = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S38=")); lggr.println(temp);
#endif
		}

		if((currentS = strstr(_commandBuffer, ("S90="))) != NULL)
		{
			offset =4;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 4, offset - 4);
			_isDcdInverted = atoi(temp);
#if DEBUG == 1
			lggr.print(F("S90=")); lggr.println(temp);
#endif
			print(F("SAVED S90=")); println(temp);
			EEPROM.write(S90_ADDRESS, atoi(temp));
		}

		if((currentS = strstr(_commandBuffer, ("S200="))) != NULL)
		{
			offset =5;
			while(currentS[offset] != '\0'
				&& isDigit(currentS[offset]))
			{
				offset++;
			}

			memset(temp, 0, 100);
			strncpy(temp, currentS + 5, offset - 5);
			_baudRate = atol(temp);
			setDefaultBaud(_baudRate);
#if DEBUG == 1
			lggr.print(F("S200=")); lggr.println(temp);
#endif
//			print(F("SAVED S200=")); println(temp);
//			EEPROM.write(S90_ADDRESS, atoi(temp));
		}

		for(int i=MAC_1 + 300; i<=USE_DHCP + 300; ++i)
		{
			char reg[6];
			sprintf(reg, "S%u=", i);

			if((currentS = strstr(_commandBuffer, reg)) != NULL)
			{
				int value = atoi(currentS + 5);
				EEPROM.write(i - 300, value);
				print(F("SAVED S")); print(i); print("="); println(value);
			}
		}

		for(int i=1; i<10; ++i)
		{
			char reg[6];
			sprintf(reg, ("S10%u="), i);
			if((currentS = strstr(_commandBuffer, reg)) != NULL)
			{
				int offset = 5;
				while(currentS[offset] != '\0')
				{
					offset++;
				}

				memset(temp, 0, 100);
				strncpy(temp, currentS + 5, offset - 5);
#if DEBUG == 1
				lggr.print(reg); lggr.println(temp);
#endif
				print(F("SAVED ")); print(reg); print('='); println(temp);
				writeAddressBook(ADDRESS_BOOK_START + (ADDRESS_BOOK_LENGTH * i), temp);
			}
		}
*/

		printOK();
	}
	else
	{
		printError();
	}

	strcpy(_lastCommandBuffer, _commandBuffer);
	resetCommandBuffer();
}

void ModemBase::ring()
{
    _isRinging = true;

	if(_S0_autoAnswer != 0) 
	{
		answer();
	}
	else
	{
		printResponse("2", F("RING"));

		digitalWrite(DCD, toggleCarrier(true));

		digitalWrite(RI, HIGH);
		delay(250);
		digitalWrite(RI, LOW);
	}
}

void ModemBase::connectOut()
{
	if(_verboseResponses)
	{
		_serial->print(F("CONNECT "));
	}

	if(_baudRate == 38400)
	{
		printResponse("28", F("38400"));
	}
	else if(_baudRate == 19200)
	{
		printResponse("15", F("19200"));
	}
	else if(_baudRate == 14400)
	{
		printResponse("13", F("14400"));
	}
	else if(_baudRate == 9600)
	{
		printResponse("12", F("9600"));
	}
	else if(_baudRate == 4800)
	{
		printResponse("11", F("4800"));
	}
	else if(_baudRate == 2400)
	{
		printResponse("10", F("2400"));
	}
	else if(_baudRate == 1200)
	{
		printResponse("5", F("1200"));
	}
	else
	{
		if(!_verboseResponses)    
			_serial->println('1');
		else
		{
			_serial->println(_baudRate);
		}
	}

	digitalWrite(DTR, toggleCarrier(true));

	_isConnected = true;
	_isCommandMode = false;
	_isRinging = false;
}

void ModemBase::processData()
{
	digitalWrite(RTS, LOW);
	//if(digitalRead(DCE_CTS) == HIGH) Serial.write("::DCE_CTS is high::");
	//if(digitalRead(DCE_CTS) == LOW) Serial.write("::DCE_CTS is low::");
	while(_serial->available())
	{
		if(_isCommandMode)
		{
			char inbound = toupper(_serial->read());

			if(_echoOn)  // && inbound != _S2_escapeCharacter
			{
				_serial->write(inbound);
			}

			if (inbound == S2_escapeCharacter)
			{
				_escapeCount++;
			}
			else
			{
				_escapeCount = 0;
			}

			if(_escapeCount == 3)
			{
				_escapeCount = 0;   // TODO, guard time!
				printOK();
			}

			if (inbound == S5_bsCharacter)
			{
				if(strlen(_commandBuffer) > 0)
				{
					_commandBuffer[strlen(_commandBuffer) - 1] = '\0';
				}
			}
			else if (inbound != '\r' && inbound != '\n' && inbound != S2_escapeCharacter)
			{
				_commandBuffer[strlen(_commandBuffer)] = inbound;

				if (_commandBuffer[0] == 'A' && _commandBuffer[1] == '/')
				{
					strcpy(_commandBuffer, _lastCommandBuffer);
					processCommandBuffer();
					resetCommandBuffer();  // To prevent A matching with A/ again
				}
			}
			else if(_commandBuffer[0] == 'A' && _commandBuffer[1] == 'T')
			{
				processCommandBuffer();
			}
			else
			{
				resetCommandBuffer();
			}
		}
		else
		{
			if(_isConnected)
			{
				char inbound = _serial->read();

				if(_echoOn) _serial->write(inbound);

				if (inbound == S2_escapeCharacter)    // TODO - refactor to get rid of duplication above
				{
					_escapeCount++;
				}
				else
				{
					_escapeCount = 0;
				}

				if(_escapeCount == 3)
				{
					_escapeCount = 0;
					_isCommandMode = true;   // TODO, guard time!

					printOK();
				}

				if (!_isCommandMode)
				{
					int result = _wifly->write(inbound);
				}
			}
		}
	}
	//digitalWrite(DCE_RTS, LOW);
}

void ModemBase::printOK(void)
{
	printResponse("0", F("OK"));
}

void ModemBase::printError(void)
{
	printResponse("4", F("ERROR"));
}


void ModemBase::printResponse(const char* code, const __FlashStringHelper * msg)
{
	_serial->println();

	if(!_verboseResponses)
		_serial->println(code);
	else
		_serial->println(msg);
}

void ModemBase::answer()
{
	if (!_isRinging)
	{
		disconnect();  // This prints "NO CARRIER"
		return;
	}

	_isConnected = true;
	_isCommandMode = false;
	_isRinging = false;

	if (_baudRate == 38400)
	{
		printResponse("28", F("CONNECT 38400"));
	}
	else if (_baudRate == 19200)
	{
		printResponse("14", F("CONNECT 19200"));
	}
	else if (_baudRate == 14400)
	{
		printResponse("13", F("CONNECT 14400"));
	}
	else if (_baudRate == 9600)
	{
		printResponse("12", F("CONNECT 9600"));
	}
	else if (_baudRate == 4800)
	{
		printResponse("11", F("CONNECT 4800"));
	}
	else if (_baudRate == 2400)
	{
		printResponse("10", F("CONNECT 2400"));
	}
	else if (_baudRate == 1200)
	{
		printResponse("5", F("CONNECT 1200"));
	}
	else
	{
		if (!_verboseResponses)
			_serial->println('1');
		else
		{
			_serial->print(F("CONNECT "));
			_serial->println(_baudRate);
		}
	}

	digitalWrite(RTS, LOW);

	TerminalMode();
}

// ----------------------------------------------------------

void ModemBase::ShowInfo()
{
	char mac[20];
	char ip[20];
	char ssid[20];

	_wifly->getMAC(mac, 20);
	_wifly->getIP(ip, 20);
	_wifly->getSSID(ssid, 20);

	_serial->print(F("MAC Address: "));  _serial->println(mac);
	_serial->print(F("IP Address:  "));  _serial->println(ip);
	_serial->print(F("Wi-Fi SSID:  "));  _serial->println(ssid);
}

void ModemBase::TerminalMode()
{
	while (_wifly->available() != -1) // -1 means closed
	{
		while (_wifly->available() > 0)
		{
			_serial->write(_wifly->read());
		}

		while (_serial->available() > 0)
		{
			_wifly->write(_serial->read());
		}

		// Alternate check for open/closed state
		if (!_wifly->isConnected())
		{
			break;
		}
	}
	disconnect();
}